### DISCUSSION TITLE
is there a way to add custom addresses for the models?

### DISCUSSION THREAD

**Discussion by sabi3293043** [comfyanonymous/ComfyUI#72] [Q&A] [answered]
is there a way to add custom addresses for the models?

i also have automatic1111 installed and dont want to copy over everything it would be like 200gbs of models loras and stuff
and i really want to use this

**Comment by tzekeong** [comment]
maybe we can use the symbolic link?

For windows:
New-Item -ItemType SymbolicLink -target /path/to/automatic111/model -path /path/to/comfyUi/model

For Linux:
ln -sf path/to/automatic111/model /path/to/comfyUi/model

**Reply by Duemellon** [reply]
I'm not familiar with symbolic links. I'll dig into it on my side.

Or, if you could be a bit more specific as to where I'm putting in the "New-item..." thing? Is that at Win+R command prompt? anaconda? added to some config file?

**Reply by tzekeong** [reply]
for windows is using powershell.

**Reply by Duemellon** [reply]
Will this have to be typed into Powershell each time you run the GUI?
** looking into Powershell right now

**Reply by tzekeong** [reply]
the command should create a permanant link to the original model.

**Reply by Duemellon** [reply]
in case it goes wrong or needs correcting, what command should I look into to clear the New-Item?

**Reply by cgwers-2** [reply]
You might find this shell extension utility useful. It lets you create hard links to the directory with content you want another app to believe is in a different location and then place a hard link to it. It's not perfect for Comfy since a1111/invoke actually have a directory structure for models, but it does help.

**Reply by Jojo-xqt** [reply]
nice~3Q

**Comment by comfyanonymous** [comment]
Yes until I add an option to set your a1111 directory and have it auto load models/etc... from there you can use symlinks, they will work.


The ComfyUI code will search subfolders and follow symlinks so you can create a link to your model folder inside the models/checkpoints/ folder for example and it will work.

**Reply by Duemellon** [reply]
Is it possible to include this in a config-type file? a json or something, that is more intuitive for someone to go in & edit the lines?

**Comment by comfyanonymous** [comment, marked as answer]
https://github.com/comfyanonymous/ComfyUI/blob/master/extra_model_paths.yaml.example

Update the ui, copy the new ComfyUI/extra_model_paths.yaml.example to ComfyUI/extra_model_paths.yaml and edit it to set the path to your a1111 ui.

**Reply by sabi3293043** [reply]
is there a way to force load the yaml i tied everything but its load automaticly loading

**Reply by Duemellon** [reply]
for people who need a little more nudge in that direction:

The original looks like --
_a111:
    base_path: path/to/stable-diffusion-webui/
    checkpoints: models/Stable-diffusion_
    
What you change is 
    _base_path: path/to/stable-diffusion-webui/_
to become
    _base_path: **c:**/stable-diffusion-webui/_

If your stable-diffusion-webui is right off the C: drive. Mine was on the E: and I had changed my folder name to SD-WEBUI so it would be
    _base_path: e:/sd-webui/_

**Reply by comfyanonymous** [reply]
make sure you update comfyui with the update/update_comfyui_only.bat or update/update_comfyui.bat or git pull depending on which version you have.

**Reply by iMDeadalus** [reply]
Just a question, if the ComfyUI is on an external hard drive, and the drive changes every time i plug into a computer. How would you set that in the yaml. I can't hard path the c:/ is there a way to be reference within a folder regardless of the drive? And just to add i am not a coder so please be gentle thanks.

**Reply by Humanoidme** [reply]
Can two or more paths be added. Example: one path being the existing A1111 directory and another one being the ComfyUI's ?

**Reply by NPDX909** [reply]
>>> WHAT LINE NUMBER IN THE FILE DO YOU EDIT <<< PLEASE and THANK YOU there are literally THREE TIMES where "base_path" appears, use line numbers, please. Thank you.

**Comment by GrimblyGorn** [comment]
This mostly works well. I did have to change it some to accommodate the secondary location for Loras in the extension folder. I also tried adding the checkpoint folder to the vaes so the .vae.pt vaes could be included as well. But i'm not sure how or if such filtering/wildcard behavior is an option for these yaml files.

This is what i have so far:
```
a111:
    base_path: /SD/stable-diffusion-webui/

    checkpoints: models/Stable-diffusion
    configs: models/Stable-diffusion
    vae: |
        models/VAE
        models/Stable-diffusion/*.vae.pt
    loras: |
        models/Lora
        extensions/sd-webui-additional-networks/models/lora
    upscale_models: |
                  models/ESRGAN
                  models/SwinIR
    embeddings: embeddings
    controlnet: models/ControlNet
```
Other than the wildcard being invalid and that folder ignored, it works rather well.

One additional point though, that likely applies to any of these loaders. When you have 1,300+ Loras it's very slow to scroll through. The up/down keys do nothing and scrolling with the mouse wheel is very very slow for such a massive list. It would be handy if that could be improved some and possibly some keyboard keys added to those list views as well.

**Reply by cgwers-2** [reply]
Scrollbars

https://www.tutorialspoint.com/python/tk_scrollbar.htm

https://www.youtube.com/watch?v=8ijKnxkaoHE

Or: https://github.com/Akascape/CTkScrollableDropdown

An accordion tree in the drop-down to keep sub-directories collapsed until clicked would be really great, too.

**Reply by NeedsMoar** [reply]
ComfyUI-Custom-Scripts has menu-per-subdirectory for LoRA and Checkpoint loaders as well as image previews (create yourself or set them from CivitAI).  The only problem I've found with it is that any LoRAs / Checkpoints in subdirectories use the relative path as the model name and don't perform further searching for it, so if you ever reorganize you'll have to re-select all of your LoRAs from the correct paths when you load an old workflow.  

Similarly any workflow you load that expects everything to be in the root checkpoint / lora path  will need them re-selected.

**Comment by Skettalee** [comment]
yeah but what about clipvision models where do you have them stored in auto1111 to put in comfui? Also in the extra_model_paths.yaml there is now a Comfyui section to put im guessing models from another comfyui models folder. 

My folders for Stable Diffusion have gotten extremely huge. 400 GB's at this point and i would like to break things up by atleast taking all the models and placing them on another drive. But will that make the generation of things slower if it has to grab the model on another drive. And most important, yall talking about symbolic links and junctions. Could I have models on another drive and then junction them to the real location they need to be or is just ediitng that extra model folder and putting all my model directories work best?

**Comment by nothingness6** [comment]
I changed the path of yaml file, but it didn't work. Here is the line that I changed : 

base_path: P:\AI_Tools\StabilityMatrix-win-x64\Data\Models\StableDiffusion\

Is it anything wrong?

**Reply by DanHues** [reply]
running into the same problem with A1111 and comfy being installed on a secondary drive. Did you find a solution to this?

**Reply by nothingness6** [reply]
Nope

**Comment by laelhalawani** [comment]
Optimally there would be a model management library that works alongside ComfyUI and provides configs for all main uis.

**Comment by laelhalawani** [comment]
Does this support relative paths? 
Can I specify any structure for `otherUI`?

Here's my file structure
```
- image_generation
  - ComfyUI
    - extra_model_paths.yaml

  - sd-webui
  - my_models
    - checkpoints
      - sdxl1.safetensors
    - vae
    - lora
```

In yaml file I commented out everything, leaving only
```
other_ui:
   base_path: ../my_models
   checkpoints: checkpoints
   vae: vae
   loras: loras
```
I restarted the main.py, models are not being detected.

**Reply by laelhalawani** [reply]
For anyone looking for an answer, had to update the path to absolute and it worked.
And yes you can use this to map your own file structure to ComfyUI, including only what you want to add.

so in my case 

DOESN'T WORK:
```
other_ui:
   base_path: ../my_models
   checkpoints: checkpoints
   vae: vae
   loras: loras
```

WORKS FINE:
```
other_ui:
   base_path: /home/lael/image_generation/my_models
   checkpoints: checkpoints
   vae: vae
   loras: loras
```

**Comment by vootox** [comment]
Not only the models but you can network ComfyUI installs by sharing Custom Nodes
[https://www.magnopus.com/blog/sharing-models-and-custom-nodes-in-comfyui](https://www.magnopus.com/blog/sharing-models-and-custom-nodes-in-comfyui)

However, my goal is to have multiple Comfy intalls on the same system with each having a restricted set from Custom Nodes Share for specialised tasks. I'd like to heed those 'conflict' warnings when installing new Custom Nodes.  The link above helps me do it.

Currenly my setup is inefficient with posssilbe conflicting nodes. I'm trialing Forge + one ComfyUI in SM (Stabilty Matrix) which I'm hesitant to include my two other ComfyUI installs so I keep them seperate. When it comes to particularly stubborn Custom Node installs require a manual 'nudge' to succeed I know I can do it with Comfy on its own (not inside Stabilty Matrix). My dilema is I also have two ComfyUI's external to SM. It may be convenient to have all the models for everything in SM and all the Custom Nodes as a share possibly also inside SM, but I'm not sure still deciding. SM seems like a promising solution but for me I'm afraid some conflicts may be more difficult to resolve within SM so I'm happy keeping at least one ComfyUI isolated from SM while sharing common resources.
